src/NServiceBus.NHibernate/SynchronizedStorage/NHibernateAmbientTransactionSynchronizedStorageSession.cs

```diff
@@ -9,28 +10,40 @@
     [SkipWeaving]
     class NHibernateAmbientTransactionSynchronizedStorageSession : CompletableSynchronizedStorageSession, INHibernateSynchronizedStorageSession
     {
-        readonly IDbConnection connection;
-        readonly bool ownsConnection;
+        Func<IDbConnection> connectionFactory;
+        Func<IDbConnection, ISession> sessionFactory;
+        ISession session;
+        IDbConnection connection;
 
-        public NHibernateAmbientTransactionSynchronizedStorageSession(ISession session, IDbConnection connection, bool ownsConnection)
+        public NHibernateAmbientTransactionSynchronizedStorageSession(Func<IDbConnection> connectionFactory, Func<IDbConnection, ISession> sessionFactory)
         {
-            this.connection = connection;
-            this.ownsConnection = ownsConnection;
-            Session = session;
+            this.connectionFactory = connectionFactory;
+            this.sessionFactory = sessionFactory;
         }
 
-        public ISession Session { get; }
-        public void Dispose()
+        public ISession Session
         {
-            Session.Flush();
-            Session.Dispose();
-
-            if (ownsConnection)
+            get
             {
-                connection.Dispose();
+                if (connection == null)
```

Comment:
If it's an issue another approach is the not directly assign them but store them in scope first and if both succeed to then copy those references.